File: internal/apiv2/watch_together_picker.go
package apiv2

import (
	"context"
	"net/http"

	"github.com/Silo-Server/silo-server/internal/api/handlers"
	mediacatalog "github.com/Silo-Server/silo-server/internal/catalog"
	"github.com/Silo-Server/silo-server/internal/watchtogether"
)

// WatchTogetherPickerService computes the "together" rows and resolves them
// to cards the caller may see.
type WatchTogetherPickerService interface {
	CheckSuggestionRoomProof(string, int, string, string) error
	RoomPicker(context.Context, string, int, string, mediacatalog.AccessFilter) (watchtogether.PickerView, error)
}
type WatchTogetherPickerInput struct {
	RoomID    string `path:"room_id"`
	RoomToken string `header:"X-Room-Token" maxLength:"4096"`
}
type WatchTogetherPickerMember struct {
	UserID          ID       `json:"user_id"`
	ProfileID       string   `json:"profile_id"`
	DisplayName     string   `json:"display_name"`
	PositionSeconds *float64 `json:"position_seconds,omitempty"`
	DurationSeconds *float64 `json:"duration_seconds,omitempty"`
}
type WatchTogetherPickerNextUp struct {
	ContentID     ID     `json:"content_id"`
	SeasonNumber  int    `json:"season_number"`
	EpisodeNumber int    `json:"episode_number"`
	Title         string `json:"title,omitempty"`
	MemberCount   int    `json:"member_count" doc:"Members whose next episode this is"`
}
type WatchTogetherPickerEntry struct {
	Item    CatalogItem                 `json:"item"`
	Members []WatchTogetherPickerMember `json:"members" doc:"Members the row applies to"`
	NextUp  *WatchTogetherPickerNextUp  `json:"next_up,omitempty" doc:"For a series: the episode most members would play next"`
}
type WatchTogetherPicker struct {
	Members          []WatchTogetherRoomMember  `json:"members"`
	ContinueTogether []WatchTogetherPickerEntry `json:"continue_together" doc:"Items two or more connected members are mid-way through, most shared first"`
	WatchlistUnion   []WatchTogetherPickerEntry `json:"watchlist_union" doc:"Items on any connected member's watchlist, most shared first"`
}
type WatchTogetherPickerOutput struct {
	CacheControl string `header:"Cache-Control"`
	Body         WatchTogetherPicker
}

func registerWatchTogetherPicker(reg *Registry) {
	op := Operation{Operation: humaOp(http.MethodGet, Prefix+"/watch-together/rooms/{room_id}/picker", "getWatchTogetherRoomPicker", "realtime", "Read the rows the room picker leads with: what connected members are watching together and the union of their watchlists, resolved to cards the caller may see. Connected members across API servers are included."), Class: ClassProfileScoped, ServiceBacked: true}
	op.Errors = []int{409}
	Register(reg, op, func(ctx context.Context, in *WatchTogetherPickerInput) (*WatchTogetherPickerOutput, error) {
		svc := reg.deps.WatchTogetherPicker
		if svc == nil {
			return nil, unavailable("watch together")
		}
		if reg.deps.CatalogAccess == nil {
			return nil, unavailable("catalog access")
		}
		user, profile, p := viewerIdentity(ctx)
		if p != nil {
			return nil, p
		}
		if err := svc.CheckSuggestionRoomProof(in.RoomID, user, profile, in.RoomToken); err != nil {
			return nil, serviceProblem(err)
		}
		filter, err := reg.deps.CatalogAccess.ContextAccessFilter(ctx, handlers.AccessFilterOptions{})
		if err != nil {
			return nil, collectionProblem(err)
		}
		view, err := svc.RoomPicker(ctx, in.RoomID, user, profile, filter)
		if err != nil {
			return nil, suggestionProblem(err)
		}
		out := &WatchTogetherPickerOutput{CacheControl: adminLogsSocketCacheControl}
		out.Body.Members = watchTogetherMembersOf(view.Members)
		out.Body.ContinueTogether = watchTogetherPickerEntriesOf(view.ContinueTogether)
		out.Body.WatchlistUnion = watchTogetherPickerEntriesOf(view.WatchlistUnion)
		return out, nil
	})
}

func watchTogetherPickerEntriesOf(rows []watchtogether.PickerViewEntry) []WatchTogetherPickerEntry {
	out := make([]WatchTogetherPickerEntry, 0, len(rows))
	for _, row := range rows {
		if row.Item == nil {
			continue
		}
		entry := WatchTogetherPickerEntry{Item: catalogItemDetailOf(row.Item).CatalogItem, Members: make([]WatchTogetherPickerMember, 0, len(row.Members))}
		for _, m := range row.Members {
			entry.Members = append(entry.Members, WatchTogetherPickerMember{UserID: IDFromInt(int64(m.UserID)), ProfileID: m.ProfileID, DisplayName: m.DisplayName, PositionSeconds: m.PositionSeconds, DurationSeconds: m.DurationSeconds})
		}
		if n := row.NextUp; n != nil {
			entry.NextUp = &WatchTogetherPickerNextUp{ContentID: ID(n.ContentID), SeasonNumber: n.SeasonNumber, EpisodeNumber: n.EpisodeNumber, Title: n.Title, MemberCount: n.MemberCount}
		}
		out = append(out, entry)
	}
	return out
}

File: web/src/api/v2/watchTogetherPicker.ts
import type { components } from "./schema";
import {
  captureProfileRequestContext,
  isCapturedProfileAuthorityActive,
  StaleApiRequestContextError,
} from "@/api/client";
import { v2 } from "./request";

export type PickerItem = components["schemas"]["CatalogItem"];

export interface PickerMember {
  user_id: number;
  profile_id: string;
  display_name: string;
  position_seconds?: number;
  duration_seconds?: number;
}

export interface PickerNextUp {
  content_id: string;
  season_number: number;
  episode_number: number;
  title?: string;
  member_count: number;
}

export interface PickerEntry {
  item: PickerItem;
  members: PickerMember[];
  next_up?: PickerNextUp;
}

export interface RoomPickerResponse {
  members: components["schemas"]["WatchTogetherRoomMember"][];
  continue_together: PickerEntry[];
  watchlist_union: PickerEntry[];
}

function numericID(value: string): number {
  const id = Number(value);
  if (!/^[1-9][0-9]*$/.test(value) || !Number.isSafeInteger(id))
    throw new Error("Invalid member identity.");
  return id;
}

function entryOf(entry: components["schemas"]["WatchTogetherPickerEntry"]): PickerEntry {
  return {
    item: entry.item,
    members: entry.members.map((member) => ({
      user_id: numericID(member.user_id),
      profile_id: member.profile_id,
      display_name: member.display_name,
      position_seconds: member.position_seconds,
      duration_seconds: member.duration_seconds,
    })),
    next_up: entry.next_up,
  };
}

/** The rows the room picker leads with, computed server-side over connected members. */
export async function readRoomPicker(
  roomId: string,
  roomToken: string,
  authority = captureProfileRequestContext(),
): Promise<RoomPickerResponse> {
  if (!authority || !isCapturedProfileAuthorityActive(authority))
    throw new StaleApiRequestContextError();
  const result = await v2("GET /api/v2/watch-together/rooms/{room_id}/picker", {
    path: { room_id: roomId },
    headers: { "X-Room-Token": roomToken },
    profileContext: authority,
    retryAuthentication: false,
  }).catch((error: unknown) => {
    if (!isCapturedProfileAuthorityActive(authority)) throw new StaleApiRequestContextError();
    throw error;
  });
  if (!isCapturedProfileAuthorityActive(authority)) throw new StaleApiRequestContextError();
  return {
    members: result.members,
    continue_together: result.continue_together.map(entryOf),
    watchlist_union: result.watchlist_union.map(entryOf),
  };
}
